Hash tables (#2)
Ada-Developers-Academy/core-interlude: hash-tables/overview-of-hash-tables.md

# Overview of Hash Tables

## Learning Goals

- Define hash table
- Define hash function
- Describe how hash tables resolve collisions

## Introduction

> Python dictionaries are fascinating! They store key-value pairs, and, as long as we have the key, getting the value has a time complexity of O(1).

Reviewer: cut the reason here. instead develop an example with a stronger parallel between lists and dicts.
```suggestion
Python dictionaries are fascinating!
```

> This is in direct comparison to arrays/lists. We could store elements in an array, but finding specific entries in an array requires looping through potentially the entire list an O(n) operation.

Reviewer: develop a comparison between lists and dicts
my main motivation for this was that the existing text mixes the notions of quickly looking up a key verses finding a value.
arrays _do_ have O(1) key lookups, but their keys are restricted to integers. But even in a dict, if we are truly looking for a _value_ in the dict, we would still need to iterate. They _keys_ can be looked up in O(1).
I know this is what was intended, but I wanted to make sure this is an apples to apples example.
```suggestion
Let's say we had a list of names `["Mickey", "Kermit", "Garfield"]`. If we let a user enter a name as input, and wanted to check whether that name was in this list, we would need to iterate over the array and check whether the user input matched each name. This is a O(n) operation.

Note that using the `in` operator on lists only helps the readability, not the performance. Internally, it still needs to check whether the value we are looking for matches any of the values in the array, one by one!

But what if we used a dictionary? Let's store our list of names as the keys of a dictionary like `{"Mickey": 1, "Kermit": 1, "Garfield": 1}`. The values don't really matter, but are required since dictionaries store key-value pairs. Here, `1` works just as well as any other value we might pick.

By storing our "list" in this way, we can now use the `in` operator on dictionaries to ask whether the user input is a key in our dictionary in a time complexity of O(1), constant time!
```

> How do Python dictionaries achieve this speed? How do keys and values get allocated memory? Under the hood, Python dictionaries are _hash tables_.

Reviewer: some minor edits, and split into two sections to separate the question and answer for a little drama.
```suggestion
How do Python dictionaries achieve this performance? How are keys and values allocated in memory?

Under the hood, Python dictionaries are _hash tables_.
```

| ------------- | --------------------------------------------------------------------------------------------------- | ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Hash Table    | A data structure that implements an associative array, which maps unique key identifiers to values. | Hash Map         | "Hash tables have an O(1) lookup time."                                                                                                                                                                       |
| Hash Function | A function used by a hash table to map a key to an index.                                           | Hashing function | "The hash function converts a key into an index in an associative array"                                                                                                                                      |

> | Collision     | When multiple keys map to the same element in a hash table's internal array.                        | -                | "The hash function wanted to insert a value into an index where a different value already existed, which is a collision. The hash table needs to handle this collision in order to keep both pieces of data." |

Reviewer: the function doesn't want to do anything, it just returns a result. it has the effect of multiple pairs being inserted at the same location
```suggestion
| Collision     | When multiple keys map to the same element in a hash table's internal array.                        | -                | "The hash function result caused a value to be inserted into an index where a different value already existed, which is a collision. The hash table needs to handle this collision in order to keep both pieces of data." |
```

# Hash Tables

A _hash table_ is a data structure designed to store key-value pairs. They have a method to look up a value by its key (find), add a key-value pair (insert), and delete a key-value pair (remove).

> Hash tables can uniquely look up values by key with a time complexity of O(1). With our current knowledge of programming, how is it possible for a data structure to achieve this, even though searching for an item in an array using linear search has a time complexity of O(n)?

Reviewer: other structures have O(1) lookups, just with key restrictions. Also this could be read as "uniquely look up" which I'm not sure how we look up something uniquely (the adverb placement is a little ambiguous).
```suggestion
Hash tables have the ability to look up values by key with a time complexity of O(1). With our current knowledge of programming, how is it possible for a data structure to achieve this, even though searching for an item in an array using linear search has a time complexity of O(n)?
```

> Hash tables achieve their speedy lookup times of O(1) by storing their key-value pairs into an array. They take advantage of how arrays have O(1) lookup operations when given an index.

Reviewer: ```suggestion
Hash tables achieve their speedy lookup times of O(1) by storing their key-value pairs in an array. They take advantage of how arrays have O(1) lookup operations when given an index.
```

> Each hash table has an _internal array_ used for storing. When a hash table inserts a key-value pair, it stores the value into the _internal array_.

Reviewer: the entire pair gets stored in the internal array. this is necessary so that after finding the correct index for some data, the hash table still needs to perform a key equality comparison (since there could be collisions)
```suggestion
Each hash table has an _internal array_ that it uses for its storage. When a hash table inserts a key-value pair, it stores the pair in the _internal array_ at a particular position.
```

> Hash tables use _hash functions_ to determine how to add the value into the internal array. **Hash functions** convert a given key into an index for the internal array. Hash tables store the values into that internal array using the computed index, essentially storing the entire key-value pair.

Reviewer: seems a little rambly. Structure this passage to continue from the thought of the previous passage (ended on discussion of position, so start with how to determine the position, echoing the language for leading).
```suggestion
To determine the position, hash tables use _hash functions_ to convert a given key into an index in the internal array. Hash tables use this computed index as the location to store the key-value pair.
```

> Because the hash function can quickly convert a key to an array index, this operation can result in very quick lookup times.

Reviewer: Starting this section with a "because" felt off to me, so did a little restructuring.
```suggestion
Hash functions themselves are designed to be fast. And since their result is used to perform constant time array lookups, the overall lookup can also be very quick!
```

> <!-- Lucidchart link  https://www.lucidchart.com/invitations/accept/5fdcf503-7d8b-4139-94d4-795bfed27883 -->
> 
> ![Hash Function Example](../assets/hash-tables_hash-tables_hash-function.png)
> 
> <!-- Hash function image source: https://en.wikipedia.org/wiki/Hash_function -->

Reviewer: moved a little later (and no longer using lucidchart version and not directly based of the wikipedia asset)
```suggestion
```

### !callout-info

## Hash Tables Implement Associative Arrays

> One reason why hash tables are interesting is because they implement an _abstract data type_ known as an _associative array_. An _abstract data type_ is any data structure that is theoretical, and described not by its concrete implementation, but by its behavior and how it's used. An _associative array_ is a theoretical data structure that describes collections that store key-value pairs.

Reviewer: switch theoretical to be used as an adjective as it reduces a little of the emphasis on being theoretical, and instead shift the focus to the more definition-y part.
```suggestion
One reason why hash tables are interesting is because they implement an _abstract data type_ known as an _associative array_. An _abstract data type_ is any theoretical data structure that is described not by its concrete implementation, but by its behavior and how it's used. An _associative array_ is a theoretical data structure that describes collections that store key-value pairs.
```

### !end-callout

## Time Complexity

> For all three operations of inserting, finding, and deleting values, hash tables have an O(1) time complexity.

Reviewer: stipulate when this complexity occurs
```suggestion
For all three operations of inserting, finding, and deleting values, hash tables have a O(1) time complexity, which occurs if each pair in the table is stored at a distinct index.
```

| Action                        | Average Case | Worst Case |
| ----------------------------- | ------------ | ---------- |
| Finding a value by key        | O(1)         | O(n)       |
| Removing a value by key       | O(1)         | O(n)       |
| Inserting a new key and value | O(1)         | O(n)       |

> 

Reviewer: I want to know what causes the worst case too. Just hint at it here since it is discussed in more detail later.
```suggestion

The worst case occurs if all pairs are stored at the same index. As we will see, this degrades to linear search.

```

### !callout-info

## We Describe Hash Table Big O by Average Case

> Normally, when we estimate the runtime of a data structure's methods with Big O, we assume the worst case. However, a well-designed hash table will attempt to prevent the worst-case scenario from occurring. Therefore when we work with hash tables we **assume the average-case**. The likelihood of the worst case from occurring depends on how well the **hash function** is implemented.

Reviewer: ```suggestion
Normally, when we estimate the runtime of a data structure's methods with Big O, we assume the worst case. However, a well-designed hash table will attempt to prevent the worst-case scenario from occurring. Therefore when we work with hash tables we **assume the average case**. The likelihood of the worst case occurring depends on how well the **hash function** is implemented.
```

### !end-callout

## Hash Function

**Hash functions** are functions responsible for mapping a given key into an integer index in the internal array.

### Example Hash Function

Imagine that we have a hash table which holds "name"-"job title" key-value pairs. This hash table needs a hash function to transform the "name" into an index.

> 

Reviewer: move the hash function image here
```suggestion

![Applying a hashing algorithm to several names. There are 5 steps: Key, Numeric Conversion, Numeric Value, mod by Array Length, and Index. The first key is Debbie Nguyen. An arrow indicates that we perform a numeric conversion on the name, resulting in 13 for the numeric value. Another arrow indicates the mod step (mod 4 since there are 4 indices) resulting in index 1. This will end up being a collision and is shown in fuchsia. The remaining 4 sets of data are Carlton Griffith→16→0 which will be a collision, Eddie Bowers→12→0 which is a collision with Carlton Griffith, Kaye Campbell→13→1 which is a collision with Debbie Nguyen, and Jeanette Barker→15→3 which is not a collision and is shown in yellow.](../assets/hash-tables_hash-tables_hash-function.png)  
_Fig. Mapping names to indices. Here, we used `len()` to convert each name to a number, then performed a mod by 4 to find the index in the internal array, since the internal array currently has 4 slots. Collisions can happen as a result of either the initial numeric conversion, or as a result of the squeezing forced by the modulo operation._

```

An example hash function could use the following algorithm:

> - Convert the key to a string
> - Grab the length of this string
> - Get modulo of the length of the string and the length of the internal array
> - Use this modulo as the index

Reviewer: adjust verbiage to match the image a little better
```suggestion
- Convert the key to a string
- Convert the string to a numeric value by grabbing its length
- Calculate the length of the string modulo the length of the internal array
- Use this result as the index
```

```python
def hash_code(key):
  return len(str(key)) % len(hash_array)
```

### !callout-info

## Why Modulo in This Example?

> When arrays get created, they get created with a fixed size in memory. This means that they have a maximum length and _can't_ hold an infinite number of things (unless there is an operation which expands the array. Python's list method `append` manages this for us!). Using `len(str(key)) % len(hash_array)` will always result in an index between `0` and the length of the array, meaning it will always be a valid index!

Reviewer: I think it will keep things simpler to avoid merging this discussion with thinking about how arrays "grow". Let's pretend that doesn't exist. 🙃
removed a little redundancy
add an additional tidbit about how often, there is a part of the calculation handled by the key type itself, and a part handled by the table
```suggestion
When arrays get created, they have a fixed size in memory. Trying to read or write beyond the end would result in an `IndexError`. Using `len(str(key)) % len(hash_array)` will always result in an index between `0` and the length of the array, meaning it will always be a valid index!

<br />

In practice, the class representing the key type will be responsible for converting the key to some initial numeric value (Python classes can implement the `__hash__` method for this purpose), and the hash table performs the additional steps to make sure the numeric value refers to a valid index in its internal storage.
```

### !end-callout

> 

Reviewer: I can't in good conscious leave the example hash function sitting there without acknowledging how terrible it is!
```suggestion

### !callout-danger

## Production Hash Functions Are More Complicated

The hypothetical hash function presented in this example is _not_ a good hash function. Words in languages tend to cluster around certain lengths, so there will tend to be many keys with the same length, and hence the same index in the internal array after squeezing to the array length. Production hash functions tend to perform a variety of manipulations to "stir" the bits representing our data into more varied ranges.

### !end-callout

```

## Collisions

> There is an enormous number of possible keys and a limited-size array. At some point, a hash function will result in some keys being mapped to the same index. When this occurs, it is called a _collision_.

Reviewer: rephrase to reinforce the no limit v. limited-size dichotomy.
```suggestion
There is practically no limit to the number of possible keys would could use in our hash table. By contrast, it stores our pair data in a limited-size array. At some point, any hash function will result in some keys being mapped to the same index. When this occurs, it is called a _collision_.
```

> 

Reviewer: couldn't resist providing a little detour into the logical framework for why we can definitely say that collisions are inevitable. Love the pigeon hole principle!
```suggestion

### !callout-info

## Collisions Are Predicted by the Pigeon Hole Principle

Any time we try to put `n + 1` items into `n` positions, we will be unable to put each item into a unique position. If we had only `n` items, each item could go into each of the `n` positions perfectly. But as soon as we add one more item, it must now go into the same position as a previous item! This observation is referred to as the _pigeon hole principle_ and can be useful for thinking about the performance of certain algorithms.

### !end-callout

```

### Example Collision

Let's use the same hash function from before:

```python
def hash_code(key):
  return len(str(key)) % len(hash_array)
```

Imagine a hash table that stores "name"-"job title" key-value pairs. The hash table's internal `hash_array` has a length of `15`. Imagine inserting the following pairs, which have the following indices after applying the above hash function:

| Key (Name) | Value (Job Title)     | Index (`hash_code(key)`) |
| ---------- | --------------------- | ------------------------ |

> | `"dan"`    | `"Lead Instructor"`   | `3`                      |
> | `"chris"`  | `"CS Fun Instructor"` | `5`                      |
> | `"dee"`    | `"Lead Instructor"`   | `3`                      |
> 
> If we insert Dan first, and then eventually Dee, our hash function will attempt to add Dee at the index `3`, even though Dan already exists there.

Reviewer: update the examples
```suggestion
| `"Nora"`   | `"Devops Engineer"`   | `4`                      |
| `"Aubrey"` | `"Web Developer"`     | `6`                      |
| `"Kirk"`   | `"Systems Analyst"`   | `4`                      |

If we insert Nora first, and then eventually Kirk, our hash function will attempt to add Kirk at the index `4`, even though Nora already exists there.
```

### Resolving Collisions

> There are many ways to resolve a collision! One way is to make each element of the hash array a collection itself, and then store new elements into the collection.

Reviewer: distinguish between the plain internal array and the extra collection
```suggestion
There are many ways to resolve a collision! One way is to make each element of the hash array a collection itself, and then store new elements in the collection rather than directly in the hash array.
```

### Example Resolution

> If the values are stored in an array themselves, we can end up with a hash which looks like this:

Reviewer: clarify which array/collection
```suggestion
If our elements are stored in these supplemental collections, we could end up with a hash which looks like this:
```

> ![Bad hash function example](../assets/hash-tables_hash-tables_collision.png)

Reviewer: updated image
```suggestion
![The internal array in a hash table. Indices 0 through 7 are shown, all of which contain None, other than indices 4 and 6. Index 4 refers to a list with two key-value pairs: Nora - a Devops Engineer, and Kirk - a Systems Analyst. Index 6 refers to a list with one key-value pair: Aubrey - a Web Developer. Indices 8 through 14 are not shown, but can be presumed to all contain None.](../assets/hash-tables_hash-tables_collision.png)  
_Fig. The internal array of the hash table for the data given above, with collisions resolved by adding all pairs with the same hash value to a collection stored at the index matching the hash value._
```

In this example, we resolve collisions by creating an array at each index of the main array, then adding the key-value pair to the end of the list.

### !callout-info

## Effective Hash Functions

> What would happen if we inserted 600 items, and our hash function happened to insert all 600 items into the same key? If the hash function does not adequately spread values out over the length of the array, we'll have long chains of key-value pairs and an O(n) worst case runtime to find items.

Reviewer: the function doesn't do the insertion, it just indicates where it should go
```suggestion
What would happen if we inserted 600 items, and our hash function happened to calculate the same hash value for all 600 of them? If the hash function does not adequately spread values out over the length of the array, we'll have long chains of key-value pairs and a O(n) worst case runtime to find items.
```

> 

Reviewer: ```suggestion

<br />

```

> Effective hash functions attempt to spread all possible values over the entire data structure to avoid assigning multiple keys to the same index. Most libraries use well-proven hashing functions which normally prevent high numbers of collisions.

Reviewer: to me, prevent sounds more like the function is actively stopping something from happening, when really it's more like it's designed to do its best to avoid a situation it can't control.
```suggestion
Effective hash functions attempt to spread all possible values over the entire data structure to avoid assigning multiple keys to the same index. Most libraries use well-proven hashing functions which normally avoid high numbers of collisions.
```

### !end-callout

## Summary

> Hash tables are a data structure which allow quick lookup of a value by using a key. Each key is mapped to an index of an internal array with a specialized method called a _hash function_. The average-case lookup is O(1) and the worst-case is O(n). However, most good implementations use a _hash function_ which attempt to uniformly distribute elements and prevent collisions.

Reviewer: ```suggestion
Hash tables are a data structure which allow quick lookup of a value by using a key. Each key is mapped to an index of an internal array with a specialized method called a _hash function_. The average-case lookup is O(1) and the worst-case is O(n). However, most good implementations use a _hash function_ which attempts to uniformly distribute elements and prevent collisions.
```

Reviewer 2: Just feel that we should explicitly close with the point that we can effectively treat hash tables as O(1) for lookup etc despite their worst-case.

```suggestion
Hash tables are a data structure which allow quick lookup of a value by using a key. Each key is mapped to an index of an internal array with a specialized method called a _hash function_. The average-case lookup is O(1) and the worst-case is O(n). However, most good implementations use a _hash function_ which attempts to uniformly distribute elements and prevent collisions.

This means that for _practical purposes_ we can treat the time complexity of lookup, deletion and insertion as O(1), despite the worst-case O(n) time complexity.  This is because, the implementations we use in languages like Python, so rarely encounter worst-case scenarios.
```
